### DISCUSSION TITLE
Support multiple node query for OPC UA adapter

### DISCUSSION THREAD

**Discussion by bossenti** [apache/streampipes#890] [Ideas]
As an outcome of discussion #888, this feature request is about supporting to query multiple nodes within one OPC UA adapter.

A first idea about the possible UI feature was provided by @tenthe:
> Maybe we can have a search field that can be used to browse the nodes and then a 'results' list that contains all selected nodes.
> But I'm also not sure how we can do this generically.

More than that, we also need to brainstorm how we could solve that technically.

**Comment by Praveenstein** [comment]
So I'd suggest two options
1. Manually allow user to enter the node IDs and let it be stored as a list
2. There's a software called ua expert, it's an open-source OPC ua client using gui, it has a good feature to browse the entire server's available node IDs in a nice structured way, maybe we could have something similar and let users select the required nodes from that. (You could take a look at the ua expert is free)

**Reply by tenthe** [reply]
Thanks, I totally agree a similar feature than ua expert has would be great.
I guess the biggest question is how to integrate them.

Here is some background information: StreamPipes uses a flexible configuration mechanism that allows developers to configure user input fields with so-called StaticProperties. This has the great advantage that it is very easy to configure user input. However, if you need more complex input masks, you cannot simply add them.

Since this is such a common case, I would say that it makes sense to offer a custom solution:

- **option 1:** Use a seperate node explorer (e.g. UA expert) that can be used to get the node ids.
- **option 2:** Extend the static properties to support the described behaviour
- **option 3:** Provide a unique component for OPC UA to support the described behaviour

Are there any other ideas? (@dominikriemer do you have any thoughts on that?)

**Reply by dominikriemer** [reply]
I'd personally like a feature to browse the complete tree without having to lookup the parent node id from an external tool.
As you say, it is not straightforward to add flexible browsing behaviour to the TreeInput component, which should ideally be asynchronous to avoid loading very large trees at once.

So I see the following options currently, which @tenthe has already described:
* Allow to enter more than one nodeId as a parent node in form of  a list and show one tree per entered parent node. (pro: easy to implement, con: no ideal UX)
* Query the whole tree at once (+ add a search feature to the tree selection component) and accept that there are major performance drawbacks in case of large trees (pro: flexible + powerful, con: performance)
* Write a new StreamPipes plugin/UI view which allows asynchronous browsing and creation of adapters from this view by bypassing the StreamPipes Connect view. (very flexible + powerful, could potentially support more protocols from PLC4X, con: implementation effort)

**Reply by tenthe** [reply]
@Praveenstein do you have experience with the performance when searching the whole tree?
Especially if the OPC server is running on hardware with reduced processing power.

**Reply by Praveenstein** [reply]
When I tried with 3rd party tools(such as ua expert) there was no performance lag, but when I tried with the streampipes adaptor there was significant performance lag. This happened when I was trying to create an opc ua adaptor, when I was browsing the list of nodes to select

**Reply by Praveenstein** [reply]
But in the 3rd party tools it was kind of lazy loading, meaning the first layer of parent Ids were first visible, then when I clicked on any specific parent, it loaded it's child and so on
